File: Uber/50.pow-x-n.go
package leetcode

// MyPow50 uses binary exponentiation. Each loop square represents the next power
// of two; when the current exponent bit is 1, that power contributes to answer.
// Negative exponents invert x first.
//
// Time: O(log |n|). Space: O(1).
func MyPow50(x float64, n int) float64 {
	if n < 0 {
		x = 1 / x
		n = -n
	}
	ans := 1.0
	for n > 0 {
		if n&1 == 1 {
			ans *= x
		}
		x *= x
		n >>= 1
	}
	return ans
}

File: companies/Uber/50.pow-x-n.java
/**
 * Algorithm:
 * Fast exponentiation over the binary representation of n. Square the base each
 * step and multiply it into the answer when the current bit is 1. Negative
 * exponents invert x first.
 *
 * Java data structures:
 * long stores the exponent magnitude so Integer.MIN_VALUE can be negated.
 *
 * Complexity:
 * Time O(log |n|), space O(1).
 */
class Solution {
    public double myPow(double x, int n) {
        long power = n;
        if (power < 0) {
            x = 1.0 / x;
            power = -power;
        }
        double ans = 1.0;
        while (power != 0) {
            if ((power & 1L) == 1L) {
                ans *= x;
            }
            x *= x;
            power >>= 1;
        }
        return ans;
    }
}

